# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*I predict Random Forest Classifier will be a better prediction model*

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
X = data.drop('spam', axis=1)
y = data['spam']

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [7]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [8]:
X_train

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4576            0.00                0.0           0.00           0.0   
4401            0.00                0.0           0.00           0.0   
3707            0.17                0.0           0.17           0.0   
2362            0.00                0.0           0.00           0.0   
1537            0.00                0.0           0.00           0.0   
...              ...                ...            ...           ...   
2895            0.00                0.0           0.00           0.0   
2763            0.00                0.0           0.00           0.0   
905             0.00                0.0           0.76           0.0   
3980            0.00                0.0           0.87           0.0   
235             0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4576           0.00            0.00               0.0                 0.0   
4401           0.00            0.00               0.0                 0.0   
3707           0.00            0.00               0.0                 0.0   
2362           0.00            0.00               0.0                 0.0   
1537           2.17            0.00               0.0                 0.0   
...             ...             ...               ...                 ...   
2895           0.00            0.00               0.0                 0.0   
2763           0.00            0.00               0.0                 0.0   
905            0.76            0.00               0.5                 0.5   
3980           0.00            0.17               0.0                 0.0   
235            0.00            0.00               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4576              0.0            0.00  ...                  0.00        0.000   
4401              0.0            0.00  ...                  0.00        0.000   
3707              0.8            0.00  ...                  0.00        0.253   
2362              0.0            0.00  ...                  0.00        0.000   
1537              0.0            0.00  ...                  0.00        0.000   
...               ...             ...  ...                   ...          ...   
2895              0.0            0.00  ...                  0.00        0.000   
2763              0.0            0.00  ...                  4.76        0.000   
905               0.0            1.01  ...                  0.00        0.000   
3980              0.0            0.00  ...                  0.34        0.022   
235               0.0            0.00  ...                  0.00        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4576        0.131        0.000        0.000        0.000          0.0   
4401        0.000        0.000        0.000        0.000          0.0   
3707        0.168        0.084        0.000        0.024          0.0   
2362        0.000        0.000        0.000        0.000          0.0   
1537        0.000        0.000        0.000        0.000          0.0   
...           ...          ...          ...          ...          ...   
2895        0.000        0.000        0.000        0.000          0.0   
2763        0.000        0.000        0.000        0.000          0.0   
905         0.078        0.000        0.433        0.433          0.0   
3980        0.022        0.000        0.000        0.000          0.0   
235         0.000        0.000        0.000        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
4576                       1.488                           5   
4401                       1.571                           5   
3707                       4.665                          81   
2362                       4.228                          53   
1537                       1.333              

In [9]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler().fit(X_train)


In [10]:
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ..., -0.13974678,
        -0.23023328, -0.37548424],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.13640572,
        -0.23023328, -0.46466704],
       [ 0.21317665, -0.16744248, -0.22391235, ..., -0.01186062,
         0.13008597,  1.25168127],
       ...,
       [-0.35811925, -0.16744248,  0.94500764, ..., -0.10138496,
        -0.16385868, -0.06418577],
       [-0.35811925, -0.16744248,  1.16294188, ..., -0.1351981 ,
        -0.20178702, -0.01707033],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15524447,
        -0.23971537, -0.42933046]])

In [11]:
# Scale the training data
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.35811925, -0.16744248, -0.56071981, ...,  0.07078032,
        -0.09748408, -0.12139738],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.11736569,
        -0.14963555, -0.32836879],
       [-0.35811925,  0.24309223, -0.56071981, ..., -0.01628853,
        -0.01214531, -0.27620527],
       ...,
       [ 1.01971205, -0.16744248,  0.25158053, ..., -0.12465162,
        -0.1875639 , -0.41418621],
       [-0.35811925, -0.16744248, -0.56071981, ..., -0.15939061,
        -0.24919745, -0.47476321],
       [-0.35811925,  0.1966166 , -0.56071981, ..., -0.11410514,
        -0.20178702, -0.31154185]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [12]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression()
logistic_regression_model.fit(X_train, y_train)

print(f"Training Data Score: {logistic_regression_model.score(X_train, y_train)}")
print(f"Testing Data Score: {logistic_regression_model.score(X_test, y_test)}")


Training Data Score: 0.9217391304347826
Testing Data Score: 0.9313640312771503


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:

predictions = logistic_regression_model.predict(X_train)

results_df = pd.DataFrame({"Prediction": predictions, "Actual": y_train})

results_df

Prediction  Actual
4576           0       0
4401           1       0
3707           0       0
2362           0       0
1537           1       1
...          ...     ...
2895           0       0
2763           0       0
905            1       1
3980           0       0
235            0       1

[3450 rows x 2 columns]

In [14]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test)
# Review the predictions
results_df = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test})

results_df


Prediction  Actual
1351           0       1
1687           0       1
1297           1       1
2101           0       0
3920           0       0
...          ...     ...
1089           1       1
929            1       1
1545           0       1
4356           0       0
936            1       1

[1151 rows x 2 columns]

In [16]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.9313640312771503

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [29]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state=1, n_estimators=128)
rf_model.fit(X_train, y_train)

print(f'Training Score: {rf_model.score(X_train, y_train)}')
print(f'Testing Score: {rf_model.score(X_test, y_test)}')

Training Score: 0.9997101449275362
Testing Score: 0.9669852302345786


In [30]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_model_predictions = rf_model.predict(X_test)

# Review the predictions
print(rf_model_predictions)


[1 1 1 ... 1 0 1]


In [33]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, rf_model_predictions)


print(f"Accuracy Score : {acc_score}")

Accuracy Score : 0.9669852302345786


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

Although both model's predictions showed favorable results, the Random Forest Classifier model yielded a higher accuracy of detecting spam at
 96.7% compared to the logistic regression model's 93.1%